server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -259,16 +286,28 @@ private static List<MoveAllocationCommand> getMoveCommands(AllocationCommands co
     }
 
     private void setCurrentDesiredBalance(DesiredBalance newDesiredBalance) {
-        if (logger.isTraceEnabled()) {
-            var diff = DesiredBalance.hasChanges(currentDesiredBalance, newDesiredBalance)
-                ? "Diff: " + DesiredBalance.humanReadableDiff(currentDesiredBalance, newDesiredBalance)
-                : "No changes";
-            logger.trace("Desired balance updated: {}. {}", newDesiredBalance, diff);
+        // Update current desired balance if the old value has not been changed already by master fail-over
+        final DesiredBalance updatedDesiredBalance = currentDesiredBalanceRef.updateAndGet(current -> {
+            if (current != DesiredBalance.NOT_MASTER) {
+                return newDesiredBalance;
+            } else {
+                return current;
+            }
+        });
+
+        if (updatedDesiredBalance == newDesiredBalance) {
+            if (logger.isTraceEnabled()) {
+                var diff = DesiredBalance.hasChanges(updatedDesiredBalance, newDesiredBalance)
+                    ? "Diff: " + DesiredBalance.humanReadableDiff(updatedDesiredBalance, newDesiredBalance)
+                    : "No changes";
+                logger.trace("Desired balance updated: {}. {}", newDesiredBalance, diff);
+            } else {
+                logger.debug("Desired balance updated for [{}]", newDesiredBalance.lastConvergedIndex());
+            }
+            computedShardMovements.inc(DesiredBalance.shardMovements(updatedDesiredBalance, newDesiredBalance));
```

Comment:
Is this right? shouldn't we be using `DesiredBalance.shardMovements(previousDesiredBalabnce, newDesiredBalance)`? perhaps I'm not reading this right

Reply:
As with the diff above

Reply:
You are right. This unfortunately makes the code a bit ugly. Pushed 0b03f4edfa20e8ce954915deab9e45eac69b5d35